File: protovalidate/grpc-python/finish/invoice_server_test.py
# Copyright 2020-2025 Buf Technologies, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import datetime
import grpc
import invoice.v1.invoice_service_pb2 as invoice_service_pb2
import invoice.v1.invoice_service_pb2_grpc as invoice_service_pb2_grpc
import protovalidate
import unittest
import uuid

from concurrent import futures
from dataclasses import dataclass
from grpc_reflection.v1alpha import reflection
from grpc_status import rpc_status

from invoice.v1 import invoice_pb2
from invoice.v1.invoice_service import InvoiceService
from protovalidate.internal import field_path
from validation.interceptor import ValidationInterceptor


@dataclass
class ViolationSpec:
    constraint_id: str
    field_path: str
    message: str


class InvoiceServerTest(unittest.TestCase):
    def setUp(self):
        self.port = "50052"
        self.server = self.start_server()
        self.client = self.create_client()

    def tearDown(self):
        self.server.stop(None)

    def test_a_valid_invoice_can_be_created(self):
        req = invoice_service_pb2.CreateInvoiceRequest()
        req.invoice.CopyFrom(valid_invoice())
        response = self.client.CreateInvoice(req)
        assert hasattr(response, "invoice_id")
        assert response.invoice_id
        assert hasattr(response, "version")
        assert response.version == 1

    def test_invoice_id_must_be_a_uuid(self):
        invoice = valid_invoice()
        invoice.invoice_id = ""
        self.check_violations(
            invoice,
            [
                ViolationSpec(
                    "string.uuid_empty",
                    "invoice.invoice_id",
                    "value is empty, which is not a valid UUID",
                ),
            ],
        )

    def test_account_id_must_be_a_uuid(self):
        invoice = valid_invoice()
        invoice.account_id = ""
        self.check_violations(
            invoice,
            [
                ViolationSpec(
                    "string.uuid_empty",
                    "invoice.account_id",
                    "value is empty, which is not a valid UUID",
                ),
            ],
        )

    def test_invoice_date_must_not_have_a_time_component(self):
        invoice = valid_invoice()
        invoice.invoice_date = datetime.datetime.now()
        self.check_violations(
            invoice,
            [
                ViolationSpec(
                    "invoice_date.no_time",
                    "invoice.invoice_date",
                    "time component should be zero",
                ),
            ],
        )

    def test_at_least_one_item_must_be_provided(self):
        invoice = valid_invoice()
        while len(invoice.line_items) > 0:
            invoice.line_items.pop(0)
        self.check_violations(
            invoice,
            [
                ViolationSpec(
                    "repeated.min_items",
                    "invoice.line_items",
                    "value must contain at least 1 item(s)",
                ),
            ],
        )

    def test_two_line_items_cannot_have_the_same_identifier(self):
        invoice = valid_invoice()
        invoice.line_items[0].line_item_id = invoice.line_items[1].line_item_id
        self.check_violations(
            invoice,
            [
                ViolationSpec(
                    "line_items.unique_line_item_id",
                    "invoice.line_items",
                    "all line_item_id values must be unique",
                ),
            ],
        )

    def test_two_line_items_cannot_have_the_same_product_id_and_unit_price(self):
        invoice = valid_invoice()
        invoice.line_items[0].product_id = invoice.line_items[1].product_id
        invoice.line_items[0].unit_price = invoice.line_items[1].unit_price
        self.check_violations(
            invoice,
            [
                ViolationSpec(
                    "line_items.logically_unique",
                    "invoice.line_items",
                    "line items must be unique combinations of product_id and unit_price",
                ),
            ],
        )

    def test_line_item_a_line_item_cannot_have_the_same_line_item_id_and_product_id(
        self,
    ):
        invoice = valid_invoice()
        invoice.line_items[0].line_item_id = invoice.line_items[0].product_id
        self.check_violations(
            invoice,
            [
                ViolationSpec(
                    "line_item_id.not.product_id",
                    "invoice.line_items[0]",
                    "line_item_id can't also be a product_id",
                ),
            ],
        )

    def test_line_item_line_item_id_must_be_a_uuid(self):
        invoice = valid_invoice()
        invoice.line_items[0].line_item_id = "this is not a uuid"
        self.check_violations(
            invoice,
            [
                ViolationSpec(
                    "string.uuid",
                    "invoice.line_items[0].line_item_id",
                    "value must be a valid UUID",
                ),
            ],
        )

    def test_line_item_product_id_must_be_a_uuid(self):
        invoice = valid_invoice()
        invoice.line_items[0].product_id = "this is not a uuid"
        self.check_violations(
            invoice,
            [
                ViolationSpec(
                    "string.uuid",
                    "invoice.line_items[0].product_id",
                    "value must be a valid UUID",
                ),
            ],
        )

    def test_line_item_quantity_must_be_greater_than_0(self):
        invoice = valid_invoice()
        invoice.line_items[0].quantity = 0
        self.check_violations(
            invoice,
            [
                ViolationSpec(
                    "uint64.gt",
                    "invoice.line_items[0].quantity",
                    "value must be greater than 0",
                ),
            ],
        )

    def test_line_item_unit_price_can_be_zero(self):
        req = invoice_service_pb2.CreateInvoiceRequest()
        invoice = valid_invoice()
        invoice.line_items[0].unit_price = 0
        req.invoice.CopyFrom(valid_invoice())
        response = self.client.CreateInvoice(req)
        assert hasattr(response, "invoice_id")
        assert response.invoice_id
        assert hasattr(response, "version")
        assert response.version == 1

    def test_line_item_unit_price_can_be_nonzero(self):
        req = invoice_service_pb2.CreateInvoiceRequest()
        invoice = valid_invoice()
        invoice.line_items[0].unit_price = 1
        req.invoice.CopyFrom(valid_invoice())
        response = self.client.CreateInvoice(req)
        assert hasattr(response, "invoice_id")
        assert response.invoice_id
        assert hasattr(response, "version")
        assert response.version == 1

    def check_violations(self, invoice, violation_specs):
        all_violations = None

        try:
            req = invoice_service_pb2.CreateInvoiceRequest()
            req.invoice.CopyFrom(invoice)
            self.client.CreateInvoice(req)
        except grpc.RpcError as err:
            status = rpc_status.from_call(err)
            assert status.code == grpc.StatusCode.INVALID_ARGUMENT.value[0]
            for detail in status.details:
                assert detail.Is(protovalidate.Violations.DESCRIPTOR)
                violations = protovalidate.Violations()
                detail.Unpack(violations)
                all_violations = violations.violations

        assert all_violations is not None
        assert len(violation_specs) == len(all_violations)
        for index, spec in enumerate(violation_specs):
            violation = all_violations[index]
            assert spec.constraint_id == violation.constraint_id
            assert spec.message == violation.message
            assert spec.field_path == field_path.string(violation.field)

    def start_server(self):
        server = grpc.server(
            futures.ThreadPoolExecutor(max_workers=10),
            interceptors=(ValidationInterceptor(),),
        )

        invoice_service_pb2_grpc.add_InvoiceServiceServicer_to_server(
            InvoiceService(), server
        )
        SERVICE_NAMES = (
            invoice_service_pb2.DESCRIPTOR.services_by_name["InvoiceService"].full_name,
            reflection.SERVICE_NAME,
        )
        reflection.enable_server_reflection(SERVICE_NAMES, server)

        server.add_insecure_port("[::]:" + self.port)
        server.start()
        return server

    def create_client(self):
        channel = grpc.insecure_channel("localhost:" + self.port)
        return invoice_service_pb2_grpc.InvoiceServiceStub(channel)


def valid_invoice():
    today_start = datetime.datetime.now().replace(
        hour=0, minute=0, second=0, microsecond=0
    )

    invoice = invoice_pb2.Invoice()
    invoice.invoice_id = str(uuid.uuid4())
    invoice.account_id = str(uuid.uuid4())
    invoice.invoice_date = today_start
    line_item_0 = invoice_pb2.LineItem()
    line_item_0.line_item_id = str(uuid.uuid4())
    line_item_0.product_id = str(uuid.uuid4())
    line_item_0.quantity = 1
    line_item_0.unit_price = 1
    invoice.line_items.append(line_item_0)
    line_item_1 = invoice_pb2.LineItem()
    line_item_1.line_item_id = str(uuid.uuid4())
    line_item_1.product_id = str(uuid.uuid4())
    line_item_1.quantity = 1
    line_item_1.unit_price = 1
    invoice.line_items.append(line_item_1)
    line_item_2 = invoice_pb2.LineItem()
    line_item_2.line_item_id = str(uuid.uuid4())
    line_item_2.product_id = str(uuid.uuid4())
    line_item_2.quantity = 1
    line_item_2.unit_price = 1
    invoice.line_items.append(line_item_2)

    return invoice

File: protovalidate/grpc-java/finish/src/test/java/invoice/v1/InvoiceServerTest.java
// Copyright 2020-2025 Buf Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package invoice.v1;

import static org.junit.jupiter.api.Assertions.*;

import buf.build.example.protovalidate.ValidationInterceptor;
import build.buf.protovalidate.Validator;
import build.buf.validate.FieldPath;
import build.buf.validate.FieldPathElement;
import build.buf.validate.Violation;
import build.buf.validate.Violations;
import com.google.protobuf.InvalidProtocolBufferException;
import com.google.protobuf.Timestamp;
import com.google.rpc.Status;
import invoice.v1.gen.*;
import io.grpc.*;
import io.grpc.StatusRuntimeException;
import io.grpc.protobuf.StatusProto;
import io.grpc.protobuf.services.ProtoReflectionServiceV1;
import java.net.InetSocketAddress;
import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneOffset;
import java.time.ZonedDateTime;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.UUID;
import java.util.stream.IntStream;
import org.junit.jupiter.api.*;

public class InvoiceServerTest {
    private static class ViolationSpec {
        public final String constraintId;
        public final String fieldPath;
        public final String message;

        public ViolationSpec(String constraintId, String fieldPath, String message) {
            this.constraintId = constraintId;
            this.fieldPath = fieldPath;
            this.message = message;
        }
    }

    private static Server server;
    private static InvoiceServiceGrpc.InvoiceServiceBlockingStub invoiceClient;

    @BeforeAll
    public static void startServer() {
        ValidationInterceptor interceptor = new ValidationInterceptor(new Validator());
        ServerServiceDefinition serviceDefinition = ServerInterceptors.intercept(new InvoiceService(), interceptor);
        server = Grpc.newServerBuilderForPort(0, InsecureServerCredentials.create())
                .addService(ProtoReflectionServiceV1.newInstance())
                .addService(serviceDefinition)
                .build();

        try {
            server.start();
        } catch (Exception e) {
            throw new RuntimeException("Couldn't start server for testing", e);
        }
    }

    @BeforeAll
    public static void createClient() {
        int port = ((InetSocketAddress) server.getListenSockets().get(0)).getPort();
        ManagedChannel channel = Grpc.newChannelBuilder("localhost:" + port, InsecureChannelCredentials.create())
                .build();
        invoiceClient = InvoiceServiceGrpc.newBlockingStub(channel);
    }

    @AfterAll
    public static void shutdownServer() {
        if (server != null) {
            try {
                server.shutdown();
            } catch (Exception e) {
                throw new RuntimeException(
                        "Couldn't stop server after testing. You may need to stop it forcefully.", e);
            }
        }
    }

    @Test
    @DisplayName("A valid invoice passes validation")
    public void testValidInvoice() {
        Invoice invoice = newValidInvoice();
        CreateInvoiceRequest req =
                CreateInvoiceRequest.newBuilder().setInvoice(invoice).build();
        assertDoesNotThrow(() -> invoiceClient.createInvoice(req));
    }

    @Test
    @DisplayName("InvoiceId is required")
    public void testInvoiceIdIsRequired() {
        Invoice invoice = Invoice.newBuilder()
                .mergeFrom(newValidInvoice())
                .setInvoiceId("")
                .build();
        CreateInvoiceRequest req =
                CreateInvoiceRequest.newBuilder().setInvoice(invoice).build();

        StatusRuntimeException exception =
                assertThrows(StatusRuntimeException.class, () -> invoiceClient.createInvoice(req));
        checkStatusRuntimeException(
                exception,
                Collections.singletonList(new ViolationSpec(
                        "string.uuid_empty", "invoice.invoice_id", "value is empty, which is not a valid UUID")));
    }

    @Test
    @DisplayName("AccountId is required")
    public void testAccountIdIsRequired() {
        Invoice invoice = Invoice.newBuilder()
                .mergeFrom(newValidInvoice())
                .setAccountId("")
                .build();
        CreateInvoiceRequest req =
                CreateInvoiceRequest.newBuilder().setInvoice(invoice).build();

        StatusRuntimeException exception =
                assertThrows(StatusRuntimeException.class, () -> invoiceClient.createInvoice(req));
        checkStatusRuntimeException(
                exception,
                Collections.singletonList(new ViolationSpec(
                        "string.uuid_empty", "invoice.account_id", "value is empty, which is not a valid UUID")));
    }

    @Test
    @DisplayName("InvoiceDate may not have a time component")
    public void testInvoiceDateMayNotHaveATimeComponent() {
        Invoice invoice = Invoice.newBuilder()
                .mergeFrom(newValidInvoice())
                .setInvoiceDate(Timestamp.newBuilder()
                        .setSeconds(Instant.now().getEpochSecond())
                        .build())
                .build();
        CreateInvoiceRequest req =
                CreateInvoiceRequest.newBuilder().setInvoice(invoice).build();

        StatusRuntimeException exception =
                assertThrows(StatusRuntimeException.class, () -> invoiceClient.createInvoice(req));
        checkStatusRuntimeException(
                exception,
                Collections.singletonList(new ViolationSpec(
                        "invoice_date.no_time", "invoice.invoice_date", "time component should be zero")));
    }

    @Test
    @DisplayName("At least one line item must be provided")
    public void testAtLeastOneLineItemMustBeProvided() {
        Invoice invoice = Invoice.newBuilder()
                .mergeFrom(newValidInvoice())
                .clearLineItems()
                .build();

        CreateInvoiceRequest req =
                CreateInvoiceRequest.newBuilder().setInvoice(invoice).build();

        StatusRuntimeException exception =
                assertThrows(StatusRuntimeException.class, () -> invoiceClient.createInvoice(req));
        checkStatusRuntimeException(
                exception,
                Collections.singletonList(new ViolationSpec(
                        "repeated.min_items", "invoice.line_items", "value must contain at least 1 item(s)")));
    }

    @Test
    @DisplayName("Two line items cannot have the same identifier")
    public void testTwoLineItemsCannotHaveTheSameIdentifier() {
        Invoice template = newValidInvoice();
        Invoice invoice = Invoice.newBuilder()
                .mergeFrom(template)
                .setLineItems(
                        1,
                        LineItem.newBuilder()
                                .mergeFrom(template.getLineItems(1))
                                .setLineItemId(template.getLineItems(0).getLineItemId()))
                .build();

        CreateInvoiceRequest req =
                CreateInvoiceRequest.newBuilder().setInvoice(invoice).build();

        StatusRuntimeException exception =
                assertThrows(StatusRuntimeException.class, () -> invoiceClient.createInvoice(req));
        checkStatusRuntimeException(
                exception,
                Collections.singletonList(new ViolationSpec(
                        "line_items.unique_line_item_id",
                        "invoice.line_items",
                        "all line_item_id values must be unique")));
    }

    @Test
    @DisplayName("Two line items cannot have the same product_id and unit price")
    public void testTwoLineItemsCannotHaveTheSameProductIdAndUnitPrice() {
        Invoice template = newValidInvoice();
        Invoice invoice = Invoice.newBuilder()
                .mergeFrom(template)
                .setLineItems(
                        1,
                        LineItem.newBuilder()
                                .mergeFrom(template.getLineItems(1))
                                .setProductId(template.getLineItems(0).getProductId())
                                .setUnitPrice(template.getLineItems(0).getUnitPrice()))
                .build();

        CreateInvoiceRequest req =
                CreateInvoiceRequest.newBuilder().setInvoice(invoice).build();

        StatusRuntimeException exception =
                assertThrows(StatusRuntimeException.class, () -> invoiceClient.createInvoice(req));
        checkStatusRuntimeException(
                exception,
                Collections.singletonList(new ViolationSpec(
                        "line_items.logically_unique",
                        "invoice.line_items",
                        "line items must be unique combinations of product_id and unit_price")));
    }

    @Test
    @DisplayName("A line item cannot have the same line_item_id and product_id")
    public void testALineItemCannotHaveTheSameLineItemIdAndProductId() {
        Invoice template = newValidInvoice();
        Invoice invoice = Invoice.newBuilder()
                .mergeFrom(template)
                .setLineItems(
                        0,
                        LineItem.newBuilder()
                                .mergeFrom(template.getLineItems(0))
                                .setLineItemId(template.getLineItems(0).getProductId()))
                .build();

        CreateInvoiceRequest req =
                CreateInvoiceRequest.newBuilder().setInvoice(invoice).build();

        StatusRuntimeException exception =
                assertThrows(StatusRuntimeException.class, () -> invoiceClient.createInvoice(req));
        checkStatusRuntimeException(
                exception,
                Collections.singletonList(new ViolationSpec(
                        "line_item_id.not.product_id",
                        "invoice.line_items[0]",
                        "line_item_id can't also be a product_id")));
    }

    @Test
    @DisplayName("LineItem LineItemId must be a UUID")
    public void testLineItemLineItemIdMustBeUUID() {
        Invoice template = newValidInvoice();
        Invoice invoice = Invoice.newBuilder()
                .mergeFrom(template)
                .setLineItems(
                        0,
                        LineItem.newBuilder()
                                .mergeFrom(template.getLineItems(0))
                                .setLineItemId(""))
                .build();

        CreateInvoiceRequest req =
                CreateInvoiceRequest.newBuilder().setInvoice(invoice).build();

        StatusRuntimeException exception =
                assertThrows(StatusRuntimeException.class, () -> invoiceClient.createInvoice(req));
        checkStatusRuntimeException(
                exception,
                Collections.singletonList(new ViolationSpec(
                        "string.uuid_empty",
                        "invoice.line_items[0].line_item_id",
                        "value is empty, which is not a valid UUID")));
    }

    @Test
    @DisplayName("LineItem ProductId must be a UUID")
    public void testLineItemProductIdMustBeUUID() {
        Invoice template = newValidInvoice();
        Invoice invoice = Invoice.newBuilder()
                .mergeFrom(template)
                .setLineItems(
                        0,
                        LineItem.newBuilder()
                                .mergeFrom(template.getLineItems(0))
                                .setProductId(""))
                .build();

        CreateInvoiceRequest req =
                CreateInvoiceRequest.newBuilder().setInvoice(invoice).build();

        StatusRuntimeException exception =
                assertThrows(StatusRuntimeException.class, () -> invoiceClient.createInvoice(req));
        checkStatusRuntimeException(
                exception,
                Collections.singletonList(new ViolationSpec(
                        "string.uuid_empty",
                        "invoice.line_items[0].product_id",
                        "value is empty, which is not a valid UUID")));
    }

    @Test
    @DisplayName("LineItem Quantity must be greater than zero")
    public void testLineItemQuantityMustBeGreaterZero() {
        Invoice template = newValidInvoice();
        Invoice invoice = Invoice.newBuilder()
                .mergeFrom(template)
                .setLineItems(
                        0,
                        LineItem.newBuilder()
                                .mergeFrom(template.getLineItems(0))
                                .setQuantity(0))
                .build();

        CreateInvoiceRequest req =
                CreateInvoiceRequest.newBuilder().setInvoice(invoice).build();

        StatusRuntimeException exception =
                assertThrows(StatusRuntimeException.class, () -> invoiceClient.createInvoice(req));
        checkStatusRuntimeException(
                exception,
                Collections.singletonList(new ViolationSpec(
                        "uint64.gt", "invoice.line_items[0].quantity", "value must be greater than 0")));
    }

    @Test
    @DisplayName("LineItem UnitPrice can be zero")
    public void testLineItemUnitPriceCanBeZero() {
        Invoice template = newValidInvoice();
        Invoice invoice = Invoice.newBuilder()
                .mergeFrom(template)
                .setLineItems(
                        0,
                        LineItem.newBuilder()
                                .mergeFrom(template.getLineItems(0))
                                .setUnitPrice(0))
                .build();

        CreateInvoiceRequest req =
                CreateInvoiceRequest.newBuilder().setInvoice(invoice).build();

        assertDoesNotThrow(() -> invoiceClient.createInvoice(req));
    }

    @Test
    @DisplayName("LineItem UnitPrice can be nonzero")
    public void testLineItemUnitPriceCanBeNonzero() {
        Invoice template = newValidInvoice();
        Invoice invoice = Invoice.newBuilder()
                .mergeFrom(template)
                .setLineItems(
                        0,
                        LineItem.newBuilder()
                                .mergeFrom(template.getLineItems(0))
                                .setUnitPrice(1))
                .build();

        CreateInvoiceRequest req =
                CreateInvoiceRequest.newBuilder().setInvoice(invoice).build();

        assertDoesNotThrow(() -> invoiceClient.createInvoice(req));
    }

    private Invoice newValidInvoice() {
        LocalDate now = LocalDate.now();
        ZonedDateTime today = now.atStartOfDay(ZoneOffset.UTC);

        return Invoice.newBuilder()
                .setInvoiceId(UUID.randomUUID().toString())
                .setAccountId(UUID.randomUUID().toString())
                .setInvoiceDate(Timestamp.newBuilder()
                        .setSeconds(today.toEpochSecond())
                        .setNanos(0)
                        .build())
                .addAllLineItems(Arrays.asList(
                        LineItem.newBuilder()
                                .setLineItemId(UUID.randomUUID().toString())
                                .setProductId(UUID.randomUUID().toString())
                                .setQuantity(1L)
                                .setUnitPrice(1L)
                                .build(),
                        LineItem.newBuilder()
                                .setLineItemId(UUID.randomUUID().toString())
                                .setProductId(UUID.randomUUID().toString())
                                .setQuantity(1L)
                                .setUnitPrice(1L)
                                .build(),
                        LineItem.newBuilder()
                                .setLineItemId(UUID.randomUUID().toString())
                                .setProductId(UUID.randomUUID().toString())
                                .setQuantity(1L)
                                .setUnitPrice(1L)
                                .build()))
                .build();
    }

    private void checkStatusRuntimeException(StatusRuntimeException exception, List<ViolationSpec> violationSpecs) {
        // We should have messages from our advice
        Status status = StatusProto.fromThrowable(exception);
        assertNotNull(status);

        Violations violations = status.getDetailsList().stream()
                .filter(it -> it.is(Violations.class))
                .map(it -> {
                    try {
                        return it.unpack(Violations.class);
                    } catch (InvalidProtocolBufferException e) {
                        throw new RuntimeException("Couldn't unpack Violations", e);
                    }
                })
                .findFirst()
                .orElseThrow(() -> new RuntimeException("Couldn't unpack Violations"));

        List<Violation> violationList = violations.getViolationsList();
        assertEquals(
                violationSpecs.size(),
                violationList.size(),
                "Expected " + violationSpecs.size() + " violations, but got " + violationList.size());

        IntStream.range(0, violationSpecs.size()).forEach(i -> {
            ViolationSpec violationSpec = violationSpecs.get(i);
            Violation violation = violationList.get(i);

            assertEquals(
                    violationSpec.constraintId,
                    violation.getConstraintId(),
                    "Expected " + violationSpec.constraintId + " but got " + violation.getConstraintId());
            assertEquals(
                    violationSpec.fieldPath,
                    fieldPathString(violation.getField()),
                    "Expected " + violationSpec.fieldPath + " but got " + fieldPathString(violation.getField()));
            assertEquals(
                    violationSpec.message,
                    violation.getMessage(),
                    "Expected " + violationSpec.message + " but got " + violation.getMessage());
        });
    }

    static String fieldPathString(FieldPath fieldPath) {
        StringBuilder builder = new StringBuilder();
        for (FieldPathElement element : fieldPath.getElementsList()) {
            if (builder.length() > 0) {
                builder.append(".");
            }
            builder.append(element.getFieldName());
            switch (element.getSubscriptCase()) {
                case INDEX:
                    builder.append("[");
                    builder.append(element.getIndex());
                    builder.append("]");
                    break;
                case BOOL_KEY:
                    if (element.getBoolKey()) {
                        builder.append("[true]");
                    } else {
                        builder.append("[false]");
                    }
                    break;
                case INT_KEY:
                    builder.append("[");
                    builder.append(element.getIntKey());
                    builder.append("]");
                    break;
                case UINT_KEY:
                    builder.append("[");
                    builder.append(element.getUintKey());
                    builder.append("]");
                    break;
                case STRING_KEY:
                    builder.append("[\"");
                    builder.append(element.getStringKey().replace("\\", "\\\\").replace("\"", "\\\""));
                    builder.append("\"]");
                    break;
                case SUBSCRIPT_NOT_SET:
                    break;
            }
        }
        return builder.toString();
    }
}
